subprojects/publish/src/main/java/org/gradle/api/publish/internal/mapping/ResolutionBackedComponentDependencyResolver.java

```diff
@@ -0,0 +1,116 @@
+ */
+
+package org.gradle.api.publish.internal.mapping;
+
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.DependencyConstraint;
+import org.gradle.api.artifacts.ExternalDependency;
+import org.gradle.api.artifacts.ModuleIdentifier;
+import org.gradle.api.artifacts.ModuleVersionIdentifier;
+import org.gradle.api.artifacts.ProjectDependency;
+import org.gradle.api.artifacts.component.ComponentIdentifier;
+import org.gradle.api.artifacts.component.ModuleComponentIdentifier;
+import org.gradle.api.artifacts.result.ResolutionResult;
+import org.gradle.api.internal.artifacts.ImmutableModuleIdentifierFactory;
+import org.gradle.api.internal.artifacts.dependencies.DefaultProjectDependencyConstraint;
+import org.gradle.api.internal.artifacts.dependencies.ProjectDependencyInternal;
+import org.gradle.api.internal.artifacts.ivyservice.projectmodule.ProjectDependencyPublicationResolver;
+import org.gradle.internal.lazy.Lazy;
+import org.gradle.util.Path;
+
+import javax.annotation.Nullable;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A {@link ComponentDependencyResolver} backed by a resolution result.
+ */
+public class ResolutionBackedComponentDependencyResolver implements ComponentDependencyResolver {
+
+    private final ImmutableModuleIdentifierFactory moduleIdentifierFactory;
+    private final ProjectDependencyPublicationResolver projectDependencyResolver;
+
+    private final Lazy<Map<ModuleIdentifier, String>> resolvedComponentVersions;
+
+    public ResolutionBackedComponentDependencyResolver(
+        Configuration resolutionConfiguration,
+        ImmutableModuleIdentifierFactory moduleIdentifierFactory,
+        ProjectDependencyPublicationResolver projectDependencyResolver
+    ) {
+        this.moduleIdentifierFactory = moduleIdentifierFactory;
+        this.projectDependencyResolver = projectDependencyResolver;
+
+        this.resolvedComponentVersions = Lazy.unsafe().of(() -> getResolvedComponentVersions(resolutionConfiguration));
+    }
+
+    @Nullable
+    @Override
+    public ResolvedCoordinates resolveComponentCoordinates(ExternalDependency dependency) {
+        return resolveModuleComponentCoordinates(dependency.getGroup(), dependency.getName());
+    }
+
+    @Override
+    public ResolvedCoordinates resolveComponentCoordinates(ProjectDependency dependency) {
+        // TODO: Eventually resolved project coordinates should be made available
+        //       via the resolution result.
+        Path identityPath = ((ProjectDependencyInternal) dependency).getIdentityPath();
+        ModuleVersionIdentifier identifier = projectDependencyResolver.resolveComponent(ModuleVersionIdentifier.class, identityPath);
+        return ResolvedCoordinates.create(identifier);
+    }
+
+    @Nullable
+    @Override
+    public ResolvedCoordinates resolveComponentCoordinates(DependencyConstraint dependency) {
+        assert !(dependency instanceof DefaultProjectDependencyConstraint);
```

Comment:
💅 Should this be a `Preconditions`?  It should probably emit a message explaining the situation, too.

Reply:
The assert is just there to codify the contract of the interface.
This will catch anything that has gone horribly wrong during testing, but a `Preconditions` isn't required like we might do for a public interface